core/services/llo/evm/report_codec_common.go

```diff
@@ -21,3 +32,125 @@ func ExtractTimestamps(report llo.Report) (validAfterSeconds, observationTimesta
+	// MANDATORY
+	Type string `json:"type"`
+	// Multiplier, if provided, will be multiplied with the stream value before
+	// encoding.
+	// OPTIONAL
+	Multiplier *ubig.Big `json:"multiplier"`
+}
+
+// getNormalizedMultiplier returns the multiplier as a decimal.Decimal, defaulting
+// to 1 if the multiplier is nil.
+//
+// Negative multipliers are ok and will work as expected, flipping the sign of
+// the value.
+func (a ABIEncoder) getNormalizedMultiplier() (multiplier decimal.Decimal) {
+	if a.Multiplier == nil {
+		multiplier = decimal.NewFromInt(1)
+	} else {
+		multiplier = decimal.NewFromBigInt(a.Multiplier.ToInt(), 0)
+	}
+	return
+}
+
+func (a ABIEncoder) applyMultiplier(d decimal.Decimal) *big.Int {
+	return d.Mul(a.getNormalizedMultiplier()).BigInt()
+}
+
+// EncodePadded uses standard ABI encoding to encode the stream value, padding
+// result to 32 bytes
+func (a ABIEncoder) EncodePadded(ctx context.Context, sv llo.StreamValue) ([]byte, error) {
+	var encode interface{}
+	switch sv := sv.(type) {
+	case *llo.Decimal:
+		if sv == nil {
+			return nil, fmt.Errorf("expected non-nil *Decimal; got: %v", sv)
+		}
+		encode = a.applyMultiplier(sv.Decimal())
+	default:
+		return nil, fmt.Errorf("unhandled type; supported types are: *llo.Decimal; got: %T", sv)
+	}
+	evmEncoderConfig := fmt.Sprintf(`[{"Name":"streamValue","Type":"%s"}]`, a.Type)
+
+	codecConfig := types.CodecConfig{Configs: map[string]types.ChainCodecConfig{
+		"evm": {TypeABI: evmEncoderConfig},
+	}}
+	c, err := codec.NewCodec(codecConfig)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create codec; %w", err)
+	}
+
+	result, err := c.Encode(ctx, map[string]any{"streamValue": encode}, "evm")
+	if err != nil {
+		return nil, fmt.Errorf("failed to encode stream value %v with ABI type %q; %w", sv, a.Type, err)
+	}
+	return result, nil
+}
+
+// EncodePacked uses packed ABI encoding to encode the stream value (no padding)
+func (a ABIEncoder) EncodePacked(ctx context.Context, sv llo.StreamValue) ([]byte, error) {
+	var v *big.Int
+	switch sv := sv.(type) {
+	case *llo.Decimal:
+		if sv == nil {
+			return nil, fmt.Errorf("expected non-nil *Decimal; got: %v", sv)
+		}
+		v = a.applyMultiplier(sv.Decimal())
+	default:
+		return nil, errors.New("EncodePacked only currently supports StreamValue type of *llo.Decimal")
+	}
+	return EncodePackedBigInt(v, a.Type)
+}
+
+// regex to match Solidity integer types (uint/int) and extract bit width
+var typeRegex = regexp.MustCompile(`^(u?int)(8|16|24|32|40|48|56|64|72|80|88|96|104|112|120|128|136|144|152|160|168|176|184|192|200|208|216|224|232|240|248|256)$`)
+
+// EncodePackedBigInt converts a *big.Int to packed EVM bytes according to the Solidity type.
+// For unsigned types ("uintN"), the value must be non-negative and fit in N bits.
+// For signed types ("intN"), the function returns the two's complement representation in N bits.
+func EncodePackedBigInt(value *big.Int, typeStr string) ([]byte, error) {
```

Comment:
For reference I dug this one up - https://github.com/ethereum/go-ethereum/pull/22273/files

this was a proposal to add the encodePacked method to the go-ethereum abi package but was discarded. Is there anything from that PR that could be utilized here?

Reply:
Not really, that PR looks incomplete